pkg/kudoctl/util/k8s/k2o.go

```diff
@@ -0,0 +1,188 @@
+	if err != nil {
+		return errors.WithMessage(err, "instances")
+	}
+	_, err = k.clientset.KudoV1alpha1().PlanExecutions(vars.Namespace).List(v1.ListOptions{})
+	if err != nil {
+		return errors.WithMessage(err, "planexecutions")
+	}
+	return nil
+}
+
+// FrameworkExistsInCluster checks if a given Framework object is installed on the current k8s cluster
+func (k *K2oClient) FrameworkExistsInCluster(name string) bool {
+	framework, err := k.clientset.KudoV1alpha1().Frameworks(vars.Namespace).Get(name, v1.GetOptions{})
+	if err != nil {
+		return false
+	}
+	fmt.Printf("framework.kudo.k8s.io/%s unchanged\n", framework.Name)
+	return true
+}
+
+// AnyFrameworkVersionExistsInCluster checks if any FrameworkVersion object matches to the given Framework name
+// in the cluster
+func (k *K2oClient) AnyFrameworkVersionExistsInCluster(framework string) bool {
+	fv, err := k.clientset.KudoV1alpha1().FrameworkVersions(vars.Namespace).List(v1.ListOptions{})
+	if err != nil {
+		return false
+	}
+	if len(fv.Items) < 1 {
+		return false
+	}
+
+	var i int
+	for _, v := range fv.Items {
+		if strings.HasPrefix(v.Name, framework) {
+			i++
+		}
+	}
+	if i < 1 {
+		return false
+	}
+	fmt.Printf("frameworkversion.kudo.k8s.io/%s unchanged\n", framework)
+	return true
+}
+
+// AnyInstanceExistsInCluster checks if any FrameworkVersion object matches to the given Framework name
+// in the cluster.
+// An Instance has two identifiers:
+// 		1) Spec.FrameworkVersion.Name
+// 		spec:
+//    		frameworkVersion:
+//      		name: kafka-2.11-2.4.0
+// 		2) LabelSelector
+// 		metadata:
+//    		creationTimestamp: "2019-02-28T14:39:20Z"
+//    		generation: 1
+//    		labels:
+//      		controller-tools.k8s.io: "1.0"
+//      		framework: kafka
+// This function also just returns true if the Instance matches a specific FrameworkVersion of a Framework
+func (k *K2oClient) AnyInstanceExistsInCluster(name, version string) bool {
+	instances, err := k.clientset.KudoV1alpha1().Instances(vars.Namespace).List(v1.ListOptions{LabelSelector: "framework=" + name})
+	if err != nil {
+		return false
+	}
+	if len(instances.Items) < 1 {
+		return false
+	}
+
+	// Todo: check function that actual checks for the FrameworkVersion named e.g. "test-1.0" to exist
+	var i int
+	for _, v := range instances.Items {
+		if v.Spec.FrameworkVersion.Name == name+"-"+version {
+			i++
+		}
+	}
+
+	// This is when we don't find the version we are looking for
+	if i < 1 {
+		return false
+	}
+	fmt.Printf("instance.kudo.k8s.io/%s unchanged\n", name)
+	return true
+}
+
+// AnyFrameworkVersionInClusterOutOfSync checks if any FrameworkVersion object matches a given Framework name and
+// if not it returns false. False means that for the given Framework the most recent official FrameworkVersion
+// is not installed in the cluster or an error occurred.
+func (k *K2oClient) FrameworkVersionInClusterOutOfSync(framework, mostRecentVersion string) bool {
```

Comment:
isn't this the same as HasVersion?  If you know the most recent, you can just call that function.

Reply:
Which `hasVersion` do you mean?

It comes with a twist that `mostRecentVersion` always comes from the official repo. It compares if a version is currently installed in your cluster and if that version is matching the most recent version available in our official GitHub repo. Which if that is not the case imo the installed cluster FrameworkVersions are out of sync.

There are two cases:

#### First Case: Having a higher version installed in the cluster than is available as latest version in the GithHub repo

Let's say a developer installed manually (via `kubectl apply -f` ) a testing version for Kafka with the `Spec.Version` field string set to `2.11-2.4.1`. In our official GitHub repo we have under package `0` just the version `2.11-2.4.0` available. In that example
`FrameworkVersionInClusterOutOfSync` would check if the Cluster returned a `FrameworkVersion` that follows the name pattern `kafka-*` (to see if in general a FrameworkVersion for this Framework has been installed) but also would check into the particular FrameworkVersion to make sure it has `Spec.Version` matching the same version string of the official repo package (and we are talking not about the package version like `0` but as in our Kafka example about the latest `package` version which is `0` and in it the `kafka-frameworkversion.yaml` with `Spec.Version` set to the string [2.11-2.4.0](https://github.com/kudobuilder/frameworks/blob/2b1151eca761c0fbe61474ba44b0bdaa4f80a0fb/repo/stable/kafka/versions/0/kafka-frameworkversion.yaml#L10). Essentially the cluster operator is ahead of master, as each package has different version naming convention we just compare the strings to match each other but not saying that they are higher or lower.

#### Second Case: Having a lower version installed in the cluster than is available as latest version in the GithHub repo

This is the opposite of the first case but with identical behavior to the first case.

In either way if there is an `out-of-sync`-event detected the CLI gives you the option to fix that and install the most recent version to stay in sync with the official repo. If you don't wish to the CLI always gives you the option to opt out and specify the `package` version via the `--package-version` flag. Via this flag you can select a particular package version that has the right `Spec.Version` set in its `frameworkversion.yaml`.